### PR TITLE
Adding TF bootstrap for Gitlab and corresponding CICD templates

### PR DESCRIPTION
**Pull request by nmiodice**
## All Submissions:
-------------------------------------
* [YES] Have you followed the guidelines in our Contributing [document](./CONTRIBUTING.md)?
* [YES] Have you added an explanation of what your changes do and why you'd like us to include them?
* [YES] I have updated the documentation accordingly.
* [NO] I have added tests to cover my changes.
* [YES] All new and existing tests passed.
* [YES] My code follows the code style of this project.
* [YES] I ran lint checks locally prior to submission.
* [YES] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

## What is the current behavior?
-------------------------------------
There is no gitlab support

Issue Number: N/A

## What is the new behavior?
-------------------------------------
- This change introduces support for the following:
  - Template to bootstrap the initialization of Azure and Gitlab as to enable CICD out of the box
  - CICD templates for deploying Terraform based IAC on Gitlab

## Does this introduce a breaking change?
-------------------------------------
- [NO]

### CODE REVIEW

**Comment by TechnicallyWilliams** [comment]
Gitlab?! Oh, can't wait to review this! 👍

**Comment by nmiodice** [comment]
> Gitlab?! Oh, can't wait to review this! 👍

Looking forward as well! Thanks @TechnicallyWilliams . Any ETA on when you might have capacity to review?

File: infra/templates/gitlab-bootstrap-iac-cicd/README.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
+# Bootstrap for Terraform deployments through Gitlab into Azure
+
+This repository is intended to simplify setting up CICD required for the infrastructure CICD on the Gitlab platform. It aims to do the following:
```
Comment (TechnicallyWilliams) [inline infra/templates/gitlab-bootstrap-iac-cicd/README.md:3]:
```suggestion
This repository is intended to simplify setting up the Azure components that are needed for the Gitlab platform to perform Azure infrastructure CICD automation. It aims to do the following:
```
Relationship between Azure and Gitlab was not immediately understood by me.

Comment (nmiodice) [reply infra/templates/gitlab-bootstrap-iac-cicd/README.md:3]: I think the right documentation is a mix of what you have here and some other context that I did not include. I'll re-phrase, thanks for calling this out!

Comment (nmiodice) [reply infra/templates/gitlab-bootstrap-iac-cicd/README.md]: @TechnicallyWilliams what do you think about the latest phrasing?

Comment (TechnicallyWilliams) [reply infra/templates/gitlab-bootstrap-iac-cicd/README.md]: wow. Super clear. Love it.

File: infra/templates/gitlab-bootstrap-iac-cicd/azure.tf [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+resource "azurerm_storage_container" "tfstate" {
+  name                  = "tfstate-terraform-bootstrap"
+  storage_account_name  = azurerm_storage_account.ci.name
+  container_access_type = "private"
+}
+
+data "azurerm_client_config" "current" {}
+
```
Comment (TechnicallyWilliams) [inline infra/templates/gitlab-bootstrap-iac-cicd/azure.tf:30]: This file is stamping out 3 environment modules. Did you face any challenges trying to collapse this into one?

File: infra/templates/gitlab-bootstrap-iac-cicd/README.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+**Configure your environment**
+```bash
+# Required to configure variables in GitLab
+export GITLAB_TOKEN="..."
+
+# Required to scope variables to the GitLab group
+export TF_VAR_group_path="..."
+
+# Required to scope variables to the GitLab project that houses Terraform. This
+# should be in the form of $GROUP/$PROJECT_NAMEexport TF_VAR_gitlab_terraform_project_path="..."
```
Comment (TechnicallyWilliams) [inline infra/templates/gitlab-bootstrap-iac-cicd/README.md:71]:
```suggestion
# should be in the form of $GROUP/$PROJECT_NAME
export TF_VAR_gitlab_terraform_project_path="..."
```

File: infra/templates/gitlab-bootstrap-iac-cicd/README.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+#       should solve the issue
+terraform apply
+```
+
+Done!
+
+
+### Adding a new environment
+
+Now that Azure and GitLab have been configured to deploy the `TEK` Server through Terraform, you can easily configure Azure and GitLab to support new stages by using the `environment` module.
```
Comment (TechnicallyWilliams) [inline infra/templates/gitlab-bootstrap-iac-cicd/README.md:163]: Any links to explain "`TEK`" for the curious readers?

File: infra/templates/gitlab-bootstrap-iac-cicd/README.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+At a high level, this template aims to:
+
+* Deploy Azure Dependencies required for automated CICD of Terraform deployments
+* Configure variables in GitLab required for automated CICD of Terraform deployments
+* Configure dependencies for each a multistage (`dev`, `integration`, `prod`, etc...) Terraform deployment
+
+> **Note**: This template only sets up the **dependencies** needed to do a production ready infrastructure deployment, such as backend state, deployment credentials, Azure Contianer Reigstry and Gitlab variables.
+
+There are manythings deployed by this template, including:
```
Comment (TechnicallyWilliams) [inline infra/templates/gitlab-bootstrap-iac-cicd/README.md:15]:
```suggestion
There are many things deployed by this template, including:
```

File: infra/templates/gitlab-bootstrap-iac-cicd/README.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+## Usage
+
+There are a few use cases for the code in this repository. The sections below outline the usage for each of those cases
+
+### First Time Setup
+
+Among the many resources provisioned by this template is the [Backend Configuration](https://www.terraform.io/docs/backends/index.html) that hosts the [Terraform State](https://www.terraform.io/docs/state/index.html) for this template, as well as the state for each deployment.
+
+Because of this, we'll cannot have the backend state configured for the initial deployment of this template. These steps will take you through the following:
```
Comment (TechnicallyWilliams) [inline infra/templates/gitlab-bootstrap-iac-cicd/README.md:44]:
```suggestion
Because of this, we cannot have the backend state configured for the initial deployment of this template. These steps will take you through the following:
```

File: infra/templates/gitlab-bootstrap-iac-cicd/README.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+Now that Azure and GitLab have been configured to support the Terraform deployment, you will need to do the following to actually deploy the environment.
+
+**Trigger IAC Pipeline**
+
+You are now ready to kick off a deployment of the IAC pipeline! You can do this through the GitLab UI.
+
+### Rotate Service Principal Passwords
+
+If the need arrises to rotate the credentials for any of the generated service principals, the following command can be used to quickly rotate the credentials and also update all configuration in GitLab:
```
Comment (TechnicallyWilliams) [inline infra/templates/gitlab-bootstrap-iac-cicd/README.md:139]:
```suggestion
If the need arises to rotate the credentials for any of the generated service principals, the following command can be used to quickly rotate the credentials and also update all configuration in GitLab:
```

File: infra/templates/gitlab-bootstrap-iac-cicd/README.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+```
+
+Done!
+
+
+### Adding a new environment
+
+Now that Azure and GitLab have been configured to deploy the `TEK` Server through Terraform, you can easily configure Azure and GitLab to support new stages by using the `environment` module.
+
+> **Note**: This will only set up Auzre and GitLab to support a new environment. The environment will need to be deployed using the infrastructure deployments project (not covered here).
```
Comment (TechnicallyWilliams) [inline infra/templates/gitlab-bootstrap-iac-cicd/README.md:165]:
```suggestion
> **Note**: This will only set up Azure and GitLab to support a new environment. The environment will need to be deployed using the infrastructure deployments project (not covered here).
```

File: devops/providers/gitlab/templates/README.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+# Build and push image
+$ az acr login -n "$ACR_NAME"
+$ docker build -f Dockerfile.sample . -t "$ACR_NAME.azurecr.io/$IMAGE:$TAG"
+$ docker push "$ACR_NAME.azurecr.io/$IMAGE:$TAG"
+```
+
+**Step 3: Configure Pipeline to Use Base Image**
+
+Now customize the base image in `.gitlab-ci.yml`:
```
Comment (TechnicallyWilliams) [inline devops/providers/gitlab/templates/README.md:62]: Just a suggestion to make the instructions more dummy proof:
```suggestion
Now insert a custom value for the base image reference/property in `.gitlab-ci.yml`:
```

File: devops/providers/gitlab/templates/README.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+$ IMAGE="..."
+$ TAG="latest"
+
+# Build and push image
+$ az acr login -n "$ACR_NAME"
+$ docker build -f Dockerfile.sample . -t "$ACR_NAME.azurecr.io/$IMAGE:$TAG"
+$ docker push "$ACR_NAME.azurecr.io/$IMAGE:$TAG"
+```
+
+**Step 3: Configure Pipeline to Use Base Image**
```
Comment (TechnicallyWilliams) [inline devops/providers/gitlab/templates/README.md:60]: To be consistent as Step 2 header explicitly mentions "Gitlab":
```suggestion
**Step 3: Configure Gitlab Pipeline to Use Base Image**
```

File: devops/providers/gitlab/templates/README.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+Now customize the base image in `.gitlab-ci.yml`:
+
+> **Note**: The use of `\$CI_REGISTRY` in the command below is intentional. The [`gitlab-bootstrap-iac-cicd`](../../../../infra/templates/gitlab-bootstrap-iac-cicd/) template will configure Gitlab configure the `CI_REGISTRY` variable to point to correct container registry.
```
Comment (TechnicallyWilliams) [inline devops/providers/gitlab/templates/README.md:64]: This **Note** made me think I had to use the `gitlab-bootstrap-iac-cicd` one more time even though step 1 already cover this. Here's a suggestion:
```suggestion
> **Note**: The use of `\$CI_REGISTRY` in the command below is intentional. When this pipeline is exercised, the value of `CI_REGISTRY` will be resolved because the [`gitlab-bootstrap-iac-cicd`](../../../../infra/templates/gitlab-bootstrap-iac-cicd/) template from step 1 configured it to point to the correct container registry.
```

File: infra/templates/gitlab-bootstrap-iac-cicd/azure.tf (hunk 2 above)
Comment (nmiodice) [reply infra/templates/gitlab-bootstrap-iac-cicd/azure.tf:30]: Nope. We could have also made the module consume a list of `environments` (i.e., `['dev', 'integration', 'prod']`) and also make all the other parameters a list, but it has some interesting behaviors that we want to avoid:

For example, in the current implementation we can simply add a new environment by copy + paste of the module block. If the the module instead took a list of environments, then the caller must be sure to add new environments only to the **end**:

i.e.,  `['dev', 'integration', 'prod', 'foo-env']` and not  `['dev', 'integration', 'foo-env', 'prod']` so that the apply operation is *create only*. The first example would achieve that, whereas the second example would teardown `prod`, before creating `foo-env` and `prod` resources.

Simply put, the interface we chose makes it much more difficult to accidentally delete resource groups that contain production deployments :)

File: infra/templates/gitlab-bootstrap-iac-cicd/README.md (hunk 4 above)
Comment (nmiodice) [reply infra/templates/gitlab-bootstrap-iac-cicd/README.md:163]: Oops... I need to remove this. It is project-specific language that I can remove.

Comment (nmiodice) [reply infra/templates/gitlab-bootstrap-iac-cicd/README.md]: Removed.

**Comment by nmiodice** [comment]
Thanks for the excellent feedback @TechnicallyWilliams! I think I've incorporated the changes.

**Review by TechnicallyWilliams — APPROVED** [review]
